python/cudf/cudf/core/dtypes.py

```diff
@@ -0,0 +1,66 @@
+import pandas as pd
+from pandas.api.extensions import ExtensionDtype
+
+import cudf
+
+
+class CategoricalDtype(ExtensionDtype):
+    def __init__(self, categories=None, ordered=None):
+        """
+        dtype similar to pd.CategoricalDtype with the categories
+        stored on the GPU.
+        """
+        self._categories = self._init_categories(categories)
+        self.ordered = ordered
+
+    @property
+    def categories(self):
+        if self._categories is None:
+            return None
+        return cudf.core.index.as_index(self._categories)
+
+    @property
+    def type(self):
+        return self._categories.dtype.type
+
+    @property
+    def name(self):
+        return "category"
+
+    @property
+    def str(self):
+        return "|O08"
+
+    def to_pandas(self):
+        if self.categories is None:
+            categories = None
+        else:
+            categories = self.categories.to_pandas()
+        return pd.CategoricalDtype(categories=categories, ordered=self.ordered)
+
+    def _init_categories(self, categories):
+        if categories is None:
+            return categories
+        if len(categories) == 0:
+            dtype = "object"
+        else:
+            dtype = None
+        return cudf.core.column.as_column(categories, dtype=dtype)
+
+    def __eq__(self, other):
+        if isinstance(other, str):
+            return other == self.name
+        elif other is self:
+            return True
+        elif not isinstance(other, self.__class__):
+            return False
+        elif self.categories is None or other.categories is None:
+            return True
+        else:
+            return (
+                self._categories.dtype == other._categories.dtype
+                and self._categories.equals(other._categories)
+            )
```

Comment:
Can we use the `.categories` property instead of reaching inside the internal attribute here?

Reply:
It's easier to use `_categories` as `self.categories` incurs the overhead of converting to an `Index` first. I don't see the problem with using an internal attribute from within the class. It is guaranteed here that  both `self` and `other` are of type `cudf.CategoricalDtype` so this won't bite us.